src/riak_cs_gc.erl

```diff
@@ -359,7 +374,52 @@ mark_manifests(RiakObject, Bucket, Key, UUIDsToMark, ManiFunction, RcPid) ->
     %% with vector clock. This allows us to do a PUT
     %% again without having to re-retrieve the object
     {ok, ManifestPbc} = riak_cs_riak_client:manifest_pbc(RcPid),
-    riak_cs_pbc:put(ManifestPbc, UpdObj, [return_body], riak_cs_config:put_gckey_timeout()).
+    riak_cs_pbc:put(ManifestPbc, UpdObj, [return_body],
+                    riak_cs_config:put_gckey_timeout()). %% <= bug: put_manifest_timeout should be used here
+
+-spec maybe_delete_small_objects([cs_uuid_and_manifest()], riak_client(), non_neg_integer()) ->
+                                        {[cs_uuid_and_manifest()], [cs_uuid()]}.
+maybe_delete_small_objects(Manifests, RcPid, Threshold) ->
+    {ok, BagId} = riak_cs_riak_client:get_manifest_bag(RcPid),
+    Self = self(),
+    FinishedFun = fun(Msg) -> Self ! {maybe_delete_small_objects, Msg} end,
+    DelFun= fun({UUID, Manifest = ?MANIFEST{state=pending_delete,
+                                            content_length=ContentLength}},
+                {Survivors, UUIDsToDelete})
+                  when ContentLength < Threshold ->
+                    %% actually this won't be scheduled :P
+                    UUIDManifest = {UUID, Manifest?MANIFEST{state=scheduled_delete}},
+                    _ = lager:debug("trying to delete ~p at ~p", [UUIDManifest, BagId]),
+                    Args = [BagId, UUIDManifest, FinishedFun,
+                            dummy_gc_key_in_sync_delete,
+                            [{cleanup_manifests, false}]],
+                    {ok, Pid} = riak_cs_delete_fsm_sup:start_delete_fsm(node(), Args),
+                    Ref = erlang:monitor(process, Pid),
+                    receive
+                        {maybe_delete_small_objects, {Pid, {ok, _}}} ->
```

Comment:
From riak_cs_delete_fsm.erl:

```
deleting({block_deleted, {ok, BlockID}, DeleterPid},
         State=#state{deleted_blocks=DeletedBlocks}) ->
    UpdState = deleting_state_update(BlockID, DeleterPid, DeletedBlocks+1, State),
    ManifestState = UpdState#state.manifest?MANIFEST.state,
    deleting_state_result(ManifestState, UpdState);
deleting({block_deleted, {error, {unsatisfied_constraint, _, BlockID}}, DeleterPid},
         State=#state{deleted_blocks=DeletedBlocks}) ->
    UpdState = deleting_state_update(BlockID, DeleterPid, DeletedBlocks, State),
    ManifestState = UpdState#state.manifest?MANIFEST.state,
    deleting_state_result(ManifestState, UpdState);
```

This shows fsm does not terminate after `unsatisfied_constraint` errors,
but continue to work *without* incrementing `DeletedBlocks`.

Counterpart GC worker code is like this:

```
handle_delete_fsm_reply({ok, {_, _, _, TotalBlocks, TotalBlocks}},
                        ?STATE{current_files=[CurrentManifest | RestManifests],
                               current_fileset=FileSet,
                               block_count=BlockCount} = State) ->
    ok = continue(),
    UpdFileSet = twop_set:del_element(CurrentManifest, FileSet),
    State?STATE{delete_fsm_pid=undefined,
                current_fileset=UpdFileSet,
                current_files=RestManifests,
                block_count=BlockCount+TotalBlocks};
handle_delete_fsm_reply({ok, {_, _, _, NumDeleted, _TotalBlocks}},
                        ?STATE{current_files=[_CurrentManifest | RestManifests],
                               block_count=BlockCount} = State) ->
    ok = continue(),
    State?STATE{delete_fsm_pid=undefined,
                current_files=RestManifests,
                block_count=BlockCount+NumDeleted};
```

Remove manifest from twop_set only if `DeletedBlocks` equals to `TotalBlocks`.